cpp/src/arrow/compute/kernels/row_encoder_internal.h

```diff
@@ -287,18 +277,188 @@ void ProcessListScalar(const ScalarType& scalar, ValidFunc&& valid_func,
+    const uint8_t* child_data_buffer_1 = arr.child_data[0].GetValues<uint8_t>(1);
+    const uint8_t* child_data_buffer_2 = arr.child_data[0].GetValues<uint8_t>(2);
+
+    int64_t i = 0;
+    IterateListArray<T>(
+      arr,
+      [&](Offset list_begin, Offset list_end) {
+        lengths[i++] += get_list_length(list_begin, list_end,
+          child_validity_buffer, child_data_buffer_1, child_data_buffer_2);
+      },
+      [&] { lengths[i++] += null_list_length; });
+  } else {
+    int32_t buffer_size;
+    ProcessListScalar<T>(
+      data.scalar_as<ScalarType>(),
+      [&](int64_t list_length, const uint8_t* child_validity_buf,
+        const uint8_t* child_data_buf_1, const uint8_t* child_data_buf_2) {
+        buffer_size = get_list_length(0, list_length,
+          child_validity_buf, child_data_buf_1, child_data_buf_2);
+      },
+      [&](){
+        buffer_size = null_list_length;
+      });
+    for (int64_t i = 0; i < batch_length; i++) {
+      lengths[i] += buffer_size;
+    }
+  }
+}
+
+// A template to encode a list type for a batch.
+// get_list_length must take (same as AddListLength)
+// - the starting offset of the list
+// - the ending offset of the list
+// - the start of the validity buffer of the nested-type
+// - the start of the nested-type's buffers[1]
+// - the start of the nested-type's buffers[2] or NULLPTR if it does not exist
+// - returns: the amount of bytes to allocate for the list; getListLength(0, 0, NULLPTR,
+//   NULLPTR, NULLPTR) must return the amount of space to allocate for a null list
+// encode_element and encode_null_element encodes elements of the lists
+// - take j, representing the element's position in the child array
+// - a pointer that points to the start of where to encode for the row
+// - the start of the nested-type's buffers[1] or NULLPTR if it does not exist
+// - the start of the nested-type's buffers[2] or NULLPTR if it does not exist
+template<typename T, typename LengthFunc, typename ElementEncoder, typename NullEncoder>
+void EncodeList(const ExecValue& data, int64_t batch_length, uint8_t** encoded_bytes,
+                LengthFunc&& get_list_length, ElementEncoder&& encode_element,
+                NullEncoder&& encode_null_element) {
+  using Offset = typename T::offset_type;
+  using ScalarType = typename TypeTraits<T>::ScalarType;
+
+  auto encode_list = [&](Offset list_begin, Offset list_end,
+    const uint8_t* child_validity_buffer, const uint8_t* child_data_buffer_1,
+    const uint8_t* child_data_buffer_2, uint8_t* encoded_ptr) {
+    const uint8_t* encoded_ptr_start = encoded_ptr;
+    Offset list_length = list_end - list_begin;
+    EncodeVarLengthHeader<Offset>(encoded_ptr, KeyEncoder::kValidByte, list_length);
+    for (Offset j = list_begin; j < list_end; j++) {
+      if (child_validity_buffer == NULLPTR ||
+        bit_util::GetBit(child_validity_buffer, j)) {
+        encode_element(j, encoded_ptr, child_data_buffer_1, child_data_buffer_2);
```

Comment:
nit: to make this logic mirror how `DecodeList` works, you could have `EncodeList` handle encoding the null byte, and `encode_element` only needs to worry about encoding the actual element